### PR TITLE
Forwarding

### PR DESCRIPTION
**Pull request by thecouchcoder**

### CODE REVIEW

File: tests/test_forward.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
+#include <gtest/gtest.h>
+#include "datapath\forwarding.h"
+#include "datapath\signals.h"
```
Comment (thirtythreeforty) [inline tests/test_forward.cpp]: Please change these two lines to use forward slashes.

File: tests/test_forward.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	//ID/EX
+	idex.read1 = 0x0002;
+	idex.read2 = 0x0004;
+	
+	//EX/ME
+	exmem.write_reg = 0x0002;
+
+	//MEM/WB
+	memwb.wb_controls.mem_to_reg = 0;
+	memwb.write_reg = 0x1234;
```
Comment (thirtythreeforty) [inline tests/test_forward.cpp]: This does not mean what you think it means.  `write_reg` holds the register number that will be written, not the data that will be written to the register.  GCC complains:

```
tests/test_forward.cpp: In member function ‘virtual void Forward_detect_EXMEM_hazard_ALUSRC1_Test::TestBody()’:
tests/test_forward.cpp:22:18: warning: large integer implicitly truncated to unsigned type [-Woverflow]
  memwb.write_reg = 0x1234;
                  ^
tests/test_forward.cpp: In member function ‘virtual void Forward_detect_EXMEM_hazard_ALUSRC2_Test::TestBody()’:
tests/test_forward.cpp:51:18: warning: large integer implicitly truncated to unsigned type [-Woverflow]
  memwb.write_reg = 0x1234;
                  ^
tests/test_forward.cpp: In member function ‘virtual void Forward_detect_MEMWB_hazard_ALUSRC1_Test::TestBody()’:
tests/test_forward.cpp:76:18: warning: large integer implicitly truncated to unsigned type [-Woverflow]
  exmem.write_reg = 0x1234;
                  ^
tests/test_forward.cpp: In member function ‘virtual void Forward_detect_MEMWB_hazard_ALUSRC2_Test::TestBody()’:
tests/test_forward.cpp:105:18: warning: large integer implicitly truncated to unsigned type [-Woverflow]
  exmem.write_reg = 0x1234;
                  ^
```

**Comment by thirtythreeforty** [comment]
I'm not sure why Travis hasn't checked this, but it will fail Valgrind when it does.  After the comments are addressed, you guys want to look at merging the master branch into this to go ahead and get the datapath integrated with these changes?

File: tests/test_forward.cpp (hunk 2 above)
Comment (aharris739) [inline tests/test_forward.cpp]: That's my fault, i was trying to use a register address that was clearly different from the small addresses that were being used for the actual test. The registers can be changed to 0 to resolve the conflict.

**Comment by thirtythreeforty** [comment]
LGTM, nice work!

**Pull request merged**